revised input syncing and added support for custom inputs

Olrosse opened the pull request:
pretty much a rewrite of the input syncing code to support custom inputs and to make steering more responsive

changed steering input filter to direct and instead added temporal smoothing to every input, seems a lot more accurate and more responsive than using the pad filter, it does however add a bit of latency to inputs that where already using direct filtering

disabled local inputs on the remote vehicles so ghost controlling is no longer possible

changed gear sync to send the current gear with any input change so if a queue is spawned late it will sync as soon as there is any input

Olrosse committed:
added support for custom inputs and several other input changes

all inputs now uses direct filtering with a smoother on the receiving end, this improves steering accuracy and responseivness a lot and still looks smooth
steering now uses the steering electric rather than the steering_input electric, this means steering assists are now also synced
made it so gears are sent with any input, makes gearing desync a lot less likely when a car is in gear before it was spawned on the receiving end

Olrosse committed:
removed local ghost inputs and fixed parking brake being stuck on

disabled local inputs on remote cars with setAllowedInputSource
increased smoothing a bit to reduce jitter more, however this adds a bit more latency
made all cached inputs start at state 0 on spawn to prevent handbrakes being stuck on

Olrosse committed:
minor changes to some comments

Olrosse committed:
simplified getInputs by using the old inputsToSend table

Olrosse committed:
more comments and renamed the value storing function

Olrosse committed:
moved recievedNewData and added fix for potential lua error

Olrosse committed:
removed unused variable

OfficialLambdax commented on lua/vehicle/extensions/BeamMP/MPInputsVE.lua:
```diff
@@ -78,13 +82,28 @@ local function applyGear(data) --TODO: add handling for mismatched gearbox types
+			end
+			state = math.floor(state * 1000) / 1000
+			if lastInputsTable[inputName] ~= state then
+				inputsToSend[inputName] = {state = state}
```
is this nesting necessary?
eg. `"steering":{"state":-0.172}`. why not just `"s":-0.172`?

Olrosse replied:
no not really, the only reason it's there is because i first included the filter as well and after removing that it was convenient to just check for table on receiving end instead of checking for every of the old names, but i can change it to just using the old names, though any custom input from mods will of course have to use their actual input name cause we would have no idea what the input name would be

OfficialLambdax replied:
`though any custom input from mods will of course have to use their actual input name cause we would have no idea what the input name would be` yes so we only want to translate the common inputs, eg steering, braking etc. Really clever idea btw to just attach custom inputs

OfficialLambdax commented on lua/vehicle/extensions/BeamMP/MPInputsVE.lua:
```diff
@@ -78,13 +82,28 @@ local function applyGear(data) --TODO: add handling for mismatched gearbox types
+		end
+	end
+
+	 -- Old sync, keeping for temporary cross compatibility --
```
How temporary? im thinking of just disabling/removing the old code and make the new match the old format

Olrosse replied:
yep can do that, it's only there so it sync across this build and the current public build, i was gonna remove it as soon as this is in public

OfficialLambdax replied:
Yes i see tho as soon as this pr is merged and a new release is build, it would automatically be shipped to everyone. So the cross compatibility is really not necessary.

OfficialLambdax commented on lua/vehicle/extensions/BeamMP/MPInputsVE.lua:
```diff
@@ -100,26 +119,103 @@ local function getInputs()
+				smoother = newTemporalSmoothing(1, 1, nil, 0),
+				currentValue = 0,
+				state = 0,
+				diffrence = 0
```
This typo in `diffrence` instead of `difference` is used multiple times in this pr, tho in `onReset()` this key is named `inputData.difference`

OfficialLambdax reviewed (changes requested):
## Problem
About `if a queue is spawned late it will sync as soon as there is any input`
Issue is this
1. Remote player spawned vehicle (i kept it in the quo)
2. They changed to second gear and pressed the clutch
3. I quod
4. I tabbed to their vehicle and could see the gear being in the second but the vehicle was stalled, no clutch input was applied

## This pr changes the packet protocol
### Current format
```
  [c] = float
  [g] = int/string
  [t] = float
  [s] = float
  [b] = float
  [p] = float
```
It only sends each key over this protocol when there has been any update to it

### This PR's format
```
  [throtte] = table
    [state] = float
  [brake] = table
    [state] = float
  [steering] = table
    [state] = float
  [clutch] = table
    [state] = float
  [parkingbrake] = table
    [state] = float
```
The transmitted contents are similiar to the old, but it keeps sending the gear index with every inputs update even when the gear has not changed

## Summary
I think the new system should match the protocol of the old as in the packets contents, aka "steering" gets translated to "s" to reduce packet size. Otherwise the pr appeared to be working as intended, even if it does mix the new with the old protocol. Even testing this with a player that had a high ping to me (~150ms) the inputs update was fluid.

OfficialLambdax commented:
### Todo
- [ ] Translate common inputs to shortend versions eg. `steering` -> `s`
- [ ] Remove nesting eg. `"s":{state:0.111}` ->`"s":0.111`
- [ ] Remove cross compatibility code
- [ ] Fix typo

Olrosse committed:
fixed typo, removed nesting and removed cross compatibility code by just using the old input names instead

also changed from sending the gear with any input to sending it every 5 seconds instead

stefan750 committed:
Changed input smoother to exponential

Olrosse committed:
updated comment

OfficialLambdax reviewed (commented):
PR and Changes work as expected, thank you @Olrosse and @stefan750.

Locked for: Will perform a more sorrow test within a competitive community to see how well it runs given varying ping and game lag. Just to see if this pr makes it worse.

Olrosse committed:
fixed inputs never reaching a state of 0 causing automatics to never shift up and the parking icon to always be on

Also removed the old difference variable cause it is no longer used

Starystars67 commented:
@OfficialLambdax any update on this?
Thanks
